File: CS/ChunkAndOptSummariesExample/ViewModel/UpdateViewModelBase.cs
using DevExpress.Mvvm;
using DataUpdatesExample;
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Threading;

namespace DataUpdatesExample
{
    public class UpdateViewModelBase : ViewModelBase
    {
        protected int dataSourceSize;
        protected int groupsCount;
        protected int collectionUpdatesPerSecond;
        protected int rowUpdatesPerSecond;
        protected bool setAllRowValuesOnUpdate;
        Random rnd = new Random();
        DispatcherTimer updateTimer = new DispatcherTimer();
        IList<Device> devices;
        public IList<Device> Devices {
            get {
                return devices;
            }
            set {
                devices = value;
                RaisePropertyChanged("Devices");
            }
        }

        public UpdateViewModelBase() {
            ConfigureSettings();
            SetUpdateTimer();
            GenerateData();
        }
        protected virtual void ConfigureSettings() {

        }
        void OnUseChunkListChanged() {
            updateTimer.Stop();
            GenerateData();
            updateTimer.Start();
        }

        void SetUpdateTimer() {
            updateTimer.Interval = new TimeSpan(0, 0, 0, 0, 1000);
            updateTimer.Tick += UpdateData;
            updateTimer.Start();
        }
        protected virtual void UpdateData(object sender, EventArgs e) {
            UpdateDataCore();
        }
        protected void UpdateDataCore() {
            ModifyCollectionCore();
            ModifyDataItemsCore();
        }
        void ModifyCollectionCore() {
            for (int i = 0; i < collectionUpdatesPerSecond; i++) {
                int index = rnd.Next(0, Devices.Count);
                if (index % 2 == 0 && Devices.Count > 0)
                    Devices.RemoveAt(index);
                else {
                    Device newDevice = new Device() { Name = "Device" + i, Group = "Group" + i % groupsCount };
                    newDevice.SetRandomSensorValues(setAllRowValuesOnUpdate);
                    Devices.Insert(index, newDevice);
                }
            }
        }
        void ModifyDataItemsCore() {
            for (int i = 0; i < rowUpdatesPerSecond; i++) {
                Devices[rnd.Next(0, Devices.Count - 1)].SetRandomSensorValues(setAllRowValuesOnUpdate);
            }
        }
        void GenerateData() {
            IList<Device> newList = CreateSourceCollection();
            Random rnd = new Random();
            for (int i = 0; i < dataSourceSize; i++) {
                Device newDevice = new Device() { Name = "Device" + i, Group = "Group" + i % groupsCount };
                newDevice.SetRandomSensorValues(setAllRowValuesOnUpdate);
                newList.Add(newDevice);
            }
            Devices = newList;
        }
        protected virtual IList<Device> CreateSourceCollection() {
            return new List<Device>();
        }
    }
}

File: VB/ChunkAndOptSummariesExample/ViewModel/UpdateViewModelBase.vb
Imports DevExpress.Mvvm
Imports System
Imports System.Collections.Generic
Imports System.Windows.Threading

Namespace DataUpdatesExample

    Public Class UpdateViewModelBase
        Inherits ViewModelBase

        Protected dataSourceSize As Integer

        Protected groupsCount As Integer

        Protected collectionUpdatesPerSecond As Integer

        Protected rowUpdatesPerSecond As Integer

        Protected setAllRowValuesOnUpdate As Boolean

        Private rnd As Random = New Random()

        Private updateTimer As DispatcherTimer = New DispatcherTimer()

        Private devicesField As IList(Of Device)

        Public Property Devices As IList(Of Device)
            Get
                Return devicesField
            End Get

            Set(ByVal value As IList(Of Device))
                devicesField = value
                RaisePropertyChanged("Devices")
            End Set
        End Property

        Public Sub New()
            ConfigureSettings()
            SetUpdateTimer()
            GenerateData()
        End Sub

        Protected Overridable Sub ConfigureSettings()
        End Sub

        Private Sub OnUseChunkListChanged()
            updateTimer.Stop()
            GenerateData()
            updateTimer.Start()
        End Sub

        Private Sub SetUpdateTimer()
            updateTimer.Interval = New TimeSpan(0, 0, 0, 0, 1000)
            AddHandler updateTimer.Tick, AddressOf UpdateData
            updateTimer.Start()
        End Sub

        Protected Overridable Sub UpdateData(ByVal sender As Object, ByVal e As EventArgs)
            UpdateDataCore()
        End Sub

        Protected Sub UpdateDataCore()
            ModifyCollectionCore()
            ModifyDataItemsCore()
        End Sub

        Private Sub ModifyCollectionCore()
            For i As Integer = 0 To collectionUpdatesPerSecond - 1
                Dim index As Integer = rnd.Next(0, Devices.Count)
                If index Mod 2 = 0 AndAlso Devices.Count > 0 Then
                    Devices.RemoveAt(index)
                Else
                    Dim newDevice As Device = New Device() With {.Name = "Device" & i, .Group = "Group" & i Mod groupsCount}
                    newDevice.SetRandomSensorValues(setAllRowValuesOnUpdate)
                    Devices.Insert(index, newDevice)
                End If
            Next
        End Sub

        Private Sub ModifyDataItemsCore()
            For i As Integer = 0 To rowUpdatesPerSecond - 1
                Devices(rnd.Next(0, Devices.Count - 1)).SetRandomSensorValues(setAllRowValuesOnUpdate)
            Next
        End Sub

        Private Sub GenerateData()
            Dim newList As IList(Of Device) = CreateSourceCollection()
            Dim rnd As Random = New Random()
            For i As Integer = 0 To dataSourceSize - 1
                Dim newDevice As Device = New Device() With {.Name = "Device" & i, .Group = "Group" & i Mod groupsCount}
                newDevice.SetRandomSensorValues(setAllRowValuesOnUpdate)
                newList.Add(newDevice)
            Next

            Devices = newList
        End Sub

        Protected Overridable Function CreateSourceCollection() As IList(Of Device)
            Return New List(Of Device)()
        End Function
    End Class
End Namespace
